Preserve initial environment for config paths

georgiastuart opened the pull request:
## Problem background

I recently ran into a scope config path issue that took me a while to understand and fix! I created a scope that contained an environment variable ending in `_ROOT` to define some cache and install paths (let's say `SPACK_CUSTOM_LOC_ROOT=<some custom path>`). While the packages and misc cache items in this scope were written to the correct location in `<some custom path>/packages` or `<some custom path>/cache/misc`, the source cache and modulefiles were written to literally `<scope definition path>/${SPACK_CUSTOM_LOC_ROOT}` during `spack install`. 

Ultimately, I decided this probably wasn't intended behavior since `spack clean -d` and `spack module lmod refresh` expanded the two bad paths correctly. I believe what happens is, since the source cache and module setup is done lazily, the `clean_environment` function executed in `install` before those paths were defined by config. I may be the only one boneheaded enough to make this mistake, but I added some debug and warning messages in this PR and also added a note on what environment variables to avoid to the documentation. This PR doesn't change any Spack behavior other than adding a few messages. 

## Changes in this PR

1. Add debug messages to print the source and misc cache paths. 
2. Add a warning message when an expanded config path contains what appears to be an unexpanded environment variable (excluding `$SSL_CERT_FILE`). 
3. Add some info about environment variables to avoid in the docs.

georgiastuart committed:
Add warnings for unexpanded env vars

Signed-off-by: Georgia Stuart <gstuart@umass.edu>

georgiastuart commented:
@spackbot help

georgiastuart commented:
@spackbot fix style

georgiastuart committed:
Fix style

Signed-off-by: Georgia Stuart <gstuart@umass.edu>

georgiastuart commented:
The remaining failed checks are codecov checks that I imagine weren't met before my PR (since I didn't add any real functionality!). Do they need to pass?

tgamblin commented:
So I think this is actually a bug, not a thing we should warn about.  If the user some variable in a config file, it *really* should expand to whatever the value was before the `spack` process started, *not* what it was after we cleaned a bunch of things to sanitize the build environment. The user really shouldn't have to understand that (but good job figuring it out!).

I *think* the best thing to do would be to freeze the environment before sanitizing it, so that user config isn't affected byany of this.  e.g., something like this in `spack.util.path`:

```diff
diff --git a/lib/spack/spack/util/path.py b/lib/spack/spack/util/path.py
index 56dcd2d0e3..9fc69e7e2e 100644
--- a/lib/spack/spack/util/path.py
+++ b/lib/spack/spack/util/path.py
@@ -15,6 +15,7 @@
 import sys
 import tempfile
 from datetime import date
+from typing import Dict, Optional

 import llnl.util.tty as tty
 from llnl.util.lang import memoized
@@ -23,6 +24,9 @@

 __all__ = ["substitute_config_variables", "substitute_path_variables", "canonicalize_path"]

+#: used by freeze_environment()
+_frozen_environment: Optional[Dict] = None
+

 def architecture():
     # break circular import
@@ -186,10 +190,28 @@ def repl(match):
     return re.sub(r"(\$\w+\b|\$\{\w+\})", repl, path)


+def freeze_environment() -> None:
+    """Snapshot the environment and save it for later variable expansion.
+
+    This allows us to properly handle config variable expansion after Spack santitizes
+    its build environemnt. Users may refer to variables in config files that are cleaned
+    by Spack's :ref:`spack.build_environment.clean_environment` function. We want them
+    to expand to what the user expects them to be, so we freeze the environment before
+    sanitizing.
+
+    """
+    _frozen_environment = os.environ.copy()
+
+
+def _expand_environment_variables(path: str) -> str:
+    env = os.environ if _frozen_environment is None else _frozen_environment
+    return re.sub(r"\$?{([^\}]+)}", lambda m: env.get(m.group(1), ""), path)
+
+
 def substitute_path_variables(path):
     """Substitute config vars, expand environment vars, expand user home."""
     path = substitute_config_variables(path)
-    path = os.path.expandvars(path)
+    path = _expand_environment_variables(path)
     path = os.path.expanduser(path)
     return path
```

And a call to `freeze_environment()` right before `clean_environment()` in `setup_package()`.  Do you want to take a stab at that?

tgamblin commented:
> The remaining failed checks are codecov checks that I imagine weren't met before my PR (since I didn't add any real functionality!). Do they need to pass?

Patch coverage really should -- it's whether the code *in the PR* is covered.  I can help you write some simple tests if you want.  The "failing" project coverage is wrong, or at least it shouldn't be failing -- see #46828.  You can just ignore that one.

georgiastuart commented:
I can take a stab at the environment preservation! I think the warning should stay, though, since it'll catch env vars that weren't set but the user expects to use. I'll check out the test coverage also.

georgiastuart committed:
Add FrozenEnvironment context manager class

Signed-off-by: Georgia Stuart <gstuart@umass.edu>

georgiastuart commented:
@tgamblin I pushed up my proposed solution. I wanted to preserve the ability to use the `os.path` functions since they're cross-platform and have better logic than I wanted to replicate, so I tried to solve this by changing `os.environ` as needed with the `FrozenEnvironment` context manager I added. It should "freeze" the env whenever an object is instanciated from the class, and I added an `INITIAL_ENVIRONMENT` var that should initialize before any environment modifications are done. Seems to work fine and I don't think it should have any unintended consequences. `canonicalize_path` should use `INITIAL_ENVIRONMENT` by default, but that can be set to `None` or a different `FrozenEnvironment` object if needed.

I'll fix up the warnings and add tests tomorrow, but wanted to push this up for review in case it wasn't going to work!

georgiastuart committed:
Change try/catch to if/else to appease the linter

Signed-off-by: Georgia Stuart <gstuart@umass.edu>

becker33 commented on lib/spack/spack/util/path.py:
```diff
@@ -233,7 +249,9 @@ def add_padding(path, length):
 
-def canonicalize_path(path, default_wd=None):
+def canonicalize_path(
+    path, default_wd=None, env: Optional[FrozenEnvironment] = INITIAL_ENVIRONMENT
```
```suggestion
    path, default_wd=None, env: Union[FrozenEnvironment, contextlib.AbstractContextManager] = spack.util.lang.nullcontext()
```
As currently implemented, INITIAL_ENVIRONMENT freezes things before Spack starts, but really it should be unfrozen until the build process starts so that Spack-managed environment variables can be used

tgamblin replied:
I'd simplify this further:

```suggestion
    path, default_wd=None, env: Optional[Dict] = None
```

Which makes it easier for the caller to provide an arbitrary set of replacement values.

becker33 commented on lib/spack/spack/util/path.py:
```diff
@@ -246,23 +264,31 @@ def canonicalize_path(path, default_wd=None):
+
+        return os.path.normpath(path)
+
+    if env:
+        with env:
+            return _canonicalize_path(path, default_wd)
+    else:
+        return _canonicalize_path(path, default_wd)
```
```suggestion
    with env:
        return _canonicalize_path(path, default_wd)
```
`env` cannot be None as currently written

becker33 replied:
Considering this, probably best not to use an internal method and just do:

```
def canonicalize_path(...):
    with env:
        filename = None
        ...

tgamblin replied:
With the context manager I was suggesting you'd want something like this:

```suggestion
    # If a dictionary was provided, use it, otherwise use the frozen environment
    # as context for substituting environment variables.
    context = spack.util.lang.nullcontext()
    replacement = env if env is not None else _frozen_environment
    if replacement is not None:
        context = replace_env(replacement)
        
    with context:
        return _canonicalize_path(path, default_wd)
```

And then we could call freeze_environment() to freeze the environment before the build starts as in my earlier example, but still keep this function flexible like @georgiastuart wants.

georgiastuart replied:
This makes sense to me! I'll commit the changes tomorrow and remove the old warning wording.

tgamblin commented on lib/spack/spack/util/environment.py:
```diff
@@ -400,6 +400,24 @@ def execute(self, env: MutableMapping[str, str]):
         env[self.name] = self.separator.join(directories)
 
 
+class FrozenEnvironment:
+    """Freezes os.environ at the time of instanciation"""
+
+    def __init__(self):
+        self._frozen_environ = os.environ.copy()
+
+    def __enter__(self):
+        self._current_environ = os.environ.copy()
+        os.environ.update(self._frozen_environ)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        os.environ.clear()
+        os.environ.update(self._current_environ)
```
`os.environ.update()` adds but doesn't clear out what's already in the environment, so this doesn't really swap between the frozen env and the pre-build env -- it ends up adding the frozen environment to the real env. You have to call `clear()`

I think it's clearer to do this with a context manager function like:

```python
@contextlib.contextmanager
def replace_environment(env: Dict) -> None:
    saved_env = os.environ.copy()
    os.environ.clear()
    os.environ.update(env)

    try:
        yield
    finally:
        os.environ.clear()
        os.environ.update(saved_env)
```

which probably should go in `spack.util.environment`.

tgamblin commented:
@georgiastuart did you want us to make the changes or are you still planning to take a stab?

georgiastuart commented:
@tgamblin Sorry, day job has been hectic! I can take a stab this week, but if you need it in ASAP then I'm happy to have someone else take over.

tgamblin commented:
Nope no rush -- though we are planning a release for next Monday or the week after. This week would likely get this change in before we cut the release.
